File: php/monosecret_php/src/CallerContext.php
<?php

declare(strict_types=1);

namespace Monosecret;

/** Caller-asserted software-integration context (Monosecret 0.20+). */
final class CallerContext
{
    public function __construct(
        public readonly string $name,
        public readonly ?string $version = null,
        public readonly ?string $operation = null,
        public readonly ?string $resource = null,
    ) {
    }

    /** @return array<string, string> */
    public function toArray(): array
    {
        return array_filter(
            [
                'name' => $this->name,
                'version' => $this->version,
                'operation' => $this->operation,
                'resource' => $this->resource,
            ],
            static fn (?string $value): bool => $value !== null,
        );
    }
}

File: dotnet/monosecret_dotnet/src/Monosecret/CallerContext.cs
using System.Text.Json.Serialization;

namespace Monosecret;

/// <summary>
/// Caller-asserted software-integration context. It is audit metadata and never
/// supplies the user access reason. Available since Monosecret 0.20.
/// </summary>
public sealed record CallerContext
{
    [JsonPropertyName("name")]
    public required string Name { get; init; }

    [JsonPropertyName("version")]
    public string? Version { get; init; }

    [JsonPropertyName("operation")]
    public string? Operation { get; init; }

    [JsonPropertyName("resource")]
    public string? Resource { get; init; }
}
